File: 1043/D/prog.cpp
#include <bits/stdc++.h>


template <class Class>
void Print(Class instance) {
    std::cout << instance << std::endl;
}

void Print() {
    std::cout << std::endl;
}


int main() {
    std::ios_base::sync_with_stdio(false);
    #ifdef LOCAL
        std::ifstream input("input.txt");
        std::ofstream output("output.txt");
    #else
        auto& input = std::cin;
        auto& output = std::cout;
    #endif
    size_t n;
    size_t m;
    input >> n >> m;
    std::vector<std::vector<uint64_t>> perm(m, std::vector<uint64_t>(n));
    std::vector<std::unordered_map<uint64_t, size_t>> maps(m);
    for (size_t index = 0; index < m; ++index) {
        for (size_t second = 0; second < n; ++second) {
            input >> perm[index][second];
            maps[index][perm[index][second]] = second;
        }
    }
    uint64_t first = 0;
    uint64_t result = 0;
    while (first < n) {
        uint64_t diff = 0;
        bool flag = true;
        uint64_t el = perm[0][first];
        while (flag) {
            ++diff;
            for (size_t index = 0; index < m; ++index) {
                if (maps[index][el] + diff >= n 
                        || perm[index][maps[index][el] + diff] != perm[0][maps[0][el] + diff])
                {
                    flag = false;
                    break;
                }
            }
            if (!flag) {
                break;
            }
        }
        result += diff * (diff + 1) / 2;
        first += diff;
    }
    output << result;
    return 0;
}

File: 1043/D/prog.py
import sys


def main(inp, out):
    n, m = list(map(int, inp.readline().split()))
    perms = []
    dicts = []
    for i in range(m):
        perm = list(map(int, inp.readline().split()))
        table = dict()
        for index, char in enumerate(perm):
            table[char] = index
        perms.append(perm)
        dicts.append(table)
    first = 0
    result = 0
    while first < n:
        indices = [dicts[i][perms[0][first]] for i in range(m)]
        diff = 0
        flag = True
        while flag:
            diff += 1
            for index in range(m):
                if indices[index] + diff >= n or perms[index][indices[index] + diff] != perms[0][indices[0] + diff]:
                    flag = False
                    break
        result += diff * (diff + 1) // 2
        first += diff
    print(result, file=out)


if __name__ == '__main__':
    if sys.platform == 'win32':
        main(sys.stdin, sys.stdout)
    else:
        with open("input.txt", 'r') as inp, open("output.txt", 'w') as out:
            main(inp, out)
